README.md
OnionPhone
==========

VOIP tool for calling over Tor

OnionPhone (OPH) is a VOIP tool for calling over Tor network which can be used as a VOIP plugin for TorChat. Call is targeted to the onion address of the recipient (its hidden service HS). The recipient can install a reverse onion connection to the originator’s HS and use a faster channel periodically resetting the slower channel to reduce overall latency. 
Also provided is the ability to switch to a direct UDP connection (with NAT traversal) after the connection is established over Tor (Tor network opposed to a SIP server requiring registration and collecting metadata). And also OnionPhone can establish a direct UDP or TCP connection to the specified port on IP-address or host. 
OPH uses a custom protocol (not RTP) with only one byte unencrypted header that can provide some obfuscation against a DPI. 

OPH provides independent level of p2p encryption and authentication which employs modern cryptographic primitives: Diffe-Hellmann key exchange on Elliptic Curve 25519 and Keccak Sponge Duplexing encryption. In the case of a call to the onion address Tor protects against MitM attacks. Also the recipient can verify identity of the originator’s onion address (only with the permission of the sender) similarly to the TorChat authentication. Otherwise possible mways of ulti-factor authentication are: 

• voice (biometrics);

• using previously shared password (with the possibility of hidden notification of enforcement); 

• using long-term public keys signed with PGP. 

OPH provides Perfect Forward Secrecy (uses a fresh key for each call) and Full Deniability for using long-term public keys of participants (as a fact and a content of the conversation): like a deniable SKEME protocol uses for initial key exchange. 

OPH can use a wide range of in-build voice codec (C source is included) from ultra low-bitrates up to high quality. The full list is: MELPE-1200, MELP-2400, CODEC2-1300, CODEC2-3200, LPC10-2400, CELP4800, AMR-4750/12200+DTX,LPC-5600+VOCODER,G723- 6400, G729-8000, GSM-HR-5600, GSM-FR-13200, GSM-EFR-12400, ILBC-13333, BV16-16000, OPUS-6000VBR, SILK-10000VBR, SPEEX-15200VBR+R. Some of them are free, some require a license (check regional laws). Implemented noise suppressors (NPP7 and SPEEX) of environment sounds and automatic mic gain control. Built-in LPC vocoder with the ability of irreversible change of voice (robot, whisper etc.). Specially designed dynamic adaptive buffer useful for smart compensation of jitter in high latency Onion environment is implemented. Radio-mode (Push-to-Talk) and voice control mode (Voice Active Detector) with generation a short signal when transmission is completed are available.

OPH:

• is a command line application which does not require installation and can be run from removable disk or True-Crypt container; 

• is fully open source, developed in pure C at the possible lowest level and carefully commented;

• statically linked, does not require additional third-party libraries and uses a minimum of system functions;

• can be compiled under Linux OS (Debian, Ubuntu etc.) using GCC or under Win32 OS (from Windows 98 up to 8) using MinGW. 

Compilation: 

Source code designed for compilation in 32-bits mode. On Linux you need ALSA headers for this to work; on Ubuntu you can install them with the following command: 'sudo apt-get install libasound2-dev'. 
For compile the source code use 'make'. Executable binaries 'oph' and 'addkey' (or with '.exe' extension for Win32  using MinGW) will be created. Integrity of cryptography library can be checked running 'test' (or 'test.exe')  in '/cr'’ subfolder.  
To clear the source use 'make clean'.
For normal functionality the OnionPhone folder must contain created binaries, default configuration files ('conf.txt', 'menu.tx' and 'audiocfg') and '/key' subfolder with default files ('contacts.txt', 'guest.sec' and 'guest').

Installation: 

the easiest way to use OnionOhine as a VOIP plugin for TorChat : 

• Step 1: Put the OnionPhone folder on the hard disk, removable media or TrueCrypt container (preferred).

• Step 2: Edit the TorChat configuration file '/torchat/bin/Tor/torrc.txt': immediately following the line

	HiddenServicePort 11009 127.0.0.1:11009
	
add the a new line

	HiddenServicePort 17447 127.0.0.1:17447
	
then run the TorChat. 

• Step 3: Right click on 'myself' icon of TorChat contact list and copy the ID to clipboard. Edit the OnionPhone configuration file 'conf.txt': specify our onion address using copied ID, for example: 
	Our_onion=gegelcy5fw7dsnsn 
Once OnionPhone is started it is ready to receive incoming and make outgoing calls. 

Usage:

• To accept an incoming call press Enter. 

• To make an outgoing call as a guest to guest (without using of personal public keys) type command: 	-Oremote_onion_address
  and press Enter then wait 10-30 sec for connecting over Tor. 
  
• To toggle continious voice transmission press Enter. Hold down / release Tab for Push-to-Talk mode. 

• To apply the voice codec from 1 to 18 use the command -Ccodec number. Smaller numbers correspond to lower bitrate codecs, greater ones correspond to higher quality. Numbers from 16 to 18 correspond to the variable bit rate codec. 

• To enable security vocoder use the command -Qmode, where mode=3 correspond "whisper" voice (preferred), modes 6-255 correspond "robot" etc). For deactivation of vocoder use the command -Q-3.

• To use the chat feature type a message and send it by pressing Enter.

• To switch to direct UDP connection use the command -S (both parties must do this).

• To return into Tor from direct UDP connection use the command -O.

• To end the call use the command -H. 

• To exit the OnioPhone use the command -X or click Esc twice for emergence exiting.

Alternatively use Up, Down, left and Right arrows to navigate in menu and apply frequently used commands quickly.
  
Detailed description of OninPhone modes, commands, cryptography etc. can be found inside the document 'oph.pdf':

http://torfone.org/download/oph.pdf

Project homepage: http://torfone.org/onionphone/

Also you can always ask the author directly (Van Gegel): torfone@ukr.net or gegelcopy@ukr.net

My TorChat ID: gegelcy5fw7dsnsn.

My PGP key available at: 

https://pgp.mit.edu/pks/lookup?search=gegelcopy%40ukr.net&op=index

http://pool.sks-keyservers.net:11371/pks/lookup?search=0xB5FA66FFEFDE1F49&fingerprint=on&hash=on&op=vindex

https://www.pgpru.com/proekt/poljzovateli?profile=gegel&getkey=8C23C721D54B3AF21EF119FBB5FA66FFEFDE1F49

Note: Alpha version is presented "as is" and has not yet been checked by anyone except the developer. Do not use in "life or death" cases.


gsm/changelog

Fri Jan 29 19:02:12 1993  Jutta Degener  (jutta at kraftbus)
	
	* Release 1.0 Patchlevel 2
	fixed L_add(0,-1) in src/add.c and inc/private.h,
	thanks to Raphael Trommer at AT&T Bell Laboratories

	* ANSI C compatibility details

Fri Oct 30 17:58:54 1992  Jutta Degener  (jutta at kraftbus)

	* Release 1.0 Patchlevel 1
	Switched uid/gid in toast's [f]chown calls.

Wed Oct 28 14:12:35 1992  Carsten Bormann  (cabo at kubus)

	* Release 1.0: released
	Copyright 1992 by Jutta Degener and Carsten Bormann, Technische
	Universitaet Berlin.  See the accompanying file "COPYRIGHT" for
	details.  THERE IS ABSOLUTELY NO WARRANTY FOR THIS SOFTWARE.


codecs.h
   ///////////////////////////////////////////////
//
// **************************
// ** ENGLISH - 14/03/2013 **
//
// Project/Software name: sponge.lib
// Author: "Van Gegel" <gegelcopy@ukr.net>
//
// THIS IS A FREE SOFTWARE  AND FOR TEST ONLY!!!
// Please do not use it in the case of life and death
// This software is released under GNU LGPL:
//
// * LGPL 3.0 <http://www.gnu.org/licenses/lgpl.html>
//
// You’re free to copy, distribute and make commercial use
// of this software under the following conditions:
//
// * You have to cite the author (and copyright owner): Van Gegel
// * You have to provide a link to the author’s Homepage: <http://torfone.org>
//
///////////////////////////////////////////////


//Set of audio codecs
typedef enum{
	CODEC_CNF=0,
	//LowBitrate
	CODEC_MELPE, //1: 1200bps, 540samp->10 bytes, 5 fpp(337.5mS)->50bytes
	CODEC_CODEC21, //2: 1400bps, 320samp->7bytes, 8 fpp(320mS)->56bytes	
	CODEC_LPC10, //3: 2400bps, 180samp->7bytes, 11 fpp(247,5mS)->77bytes
	CODEC_MELP, //4: 2400bps, 180samp->7bytes, 9 fpp(202,5mS)->63bytes
	CODEC_CODEC22, //5: 3200bps, 160samp->8bytes, 9fpp(180mS)->72bytes
	CODEC_CELP, //6: 4800bps, 240samp->18bytes, 6fpp(180mS)->108bytes
	CODEC_AMRV, //7: AMR (60-200mS, 4750-12200bps, DTX)
	CODEC_LPC, //8: 5600bps, 160samp->14bytes, 9fpp(180mS)->126 bytes
	CODEC_GSMH,//9: 5600bps, 160samp->14bytes, 8fpp(160mS)->112 bytes
	//HQ
	CODEC_G723,//10: 6400bps, 240samp->24bytes, 4fpp(120mS)->96bytes
	CODEC_G729,//11: 8000bps, 80samp->10bytes, 11fpp(110mS)->110bytes
	CODEC_GSME,//12: 12400bps, 160samp->31bytes, 4fpp(100mS)->124 bytes
	CODEC_GSM, //13: 13200bps, 160samp->33bytes, 3fpp(60mS)->99 bytes
	CODEC_ILBC,//14: 13333bps, 240samp->50bytes, 2fpp(60mS)->100bytes
	CODEC_BV16,//15: 16000bps, 40samp->10bytes, 8fpp(40mS)->80bytes
	//vbr codecs
	CODEC_OPUS, //16: 6000vbr, 480sampl->25/61 bytes, 2fpp(120mS)39-122bytes
	CODEC_SILK,//17: 10000vbr, 480samp->42->96 bytes, 1fpp(60mS)
	CODEC_SPEEX //18: 15200vbr, 320 samp->11-75bytes, 1fpp(40mS) 22-86bytes
}CodecSet;

//transmit modes
typedef enum{
    TX_NONE=0, //off
    TX_VAD, //voice activation
    TX_PTT, //push-to-talk
    TX_ON //continuous
}TxMode;


int speex_i(int squal, int svbr, int prepr, int sfpp);
int speex_e(unsigned char* buf, short* speech);
int speex_er(unsigned char* buf, short* speech);
int speex_d(short* speech, unsigned char* buf, int len);
int speex_dr(short* speech, unsigned char* buf, int len);
int speex_n(short* speech, int len);
int speex_r(short* spin, short* spout, int inlen,  int outrate);
void speex_p(int denoise, int agc);
void speex_f(void);

void amr_ini(int dtx);
void amr_fin(void);

void gsm_ini(void);
void gsm_fin(void);
void gsm_a(short *sp, unsigned char* bf);
void gsm_s(unsigned char *bf, short *sp);

void gsmhr_ini(short isDTX);
void gsmhr_fin(void);
int gsmhr_a(short *sp, unsigned char* bf);
void gsmhr_s(unsigned char *bf, short *sp);

int opus_i(void);
int opus_e(unsigned char* buf, short* speech);
int opus_d(short* speech, unsigned char* buf, int len);

void lpc_i(void);
void lpc_e(unsigned char* buf, short* speech);
void lpc_d(short* speech, unsigned char* buf);
void lpc_v(short* speech, unsigned char mode);
void lpc_f(void);

void ilbc_i(int mode);
void ilbc_f(void);
int ilbc_e(unsigned char* buf, short* speech);
int ilbc_d(short* speech, unsigned char* buf);

void bv16_i(void);
void bv16_e(unsigned char* buf, short* speech);
void bv16_d(short* speech, unsigned char* buf);

int codec_type(unsigned char* bf);
int codec_len(int cdk);
int sp_encode(short* sp, unsigned char* bf);
int sp_decode(short* sp, unsigned char* bf);

int playjit(void);
int vox(short* speech, int len, char level, char hist);
long RateChange(short *src, short *dest, long srcLen, long destRate);
void sp_init(void);
void sp_fine(void);
void set_encoder(int cd);
void get_decoder(int cd);
void get_jitter(void);
int amr_setup(char mode, char dtx, char fpp);

void switch_tx(void);
void off_tx(void);
void push_ptt(void);
void go_vad(void);
void playring();

void go_snd(unsigned char* pkt);
int do_snd(unsigned char *pkt);

void setaudio(void);

